File: data/backtracking/letter-case-permutation.cpp
#include <vector>
#include <string>

using namespace std;

class SolutionLetterCase {
    vector<string> res;
    void backtrack(string& s, int i) {
        if (i == s.length()) {
            res.push_back(s);
            return;
        }
        if (isalpha(s[i])) {
            s[i] = tolower(s[i]);
            backtrack(s, i + 1);
            s[i] = toupper(s[i]);
            backtrack(s, i + 1);
        } else {
            backtrack(s, i + 1);
        }
    }
public:
    vector<string> letterCasePermutation(string s) {
        backtrack(s, 0);
        return res;
    }
};

File: data/backtracking/letter-case-permutation.py
def letterCasePermutation(s: str) -> list[str]:
    res = []
    def backtrack(sub, i):
        if i == len(s):
            res.append(sub)
            return
        if s[i].isalpha():
            backtrack(sub + s[i].lower(), i + 1)
            backtrack(sub + s[i].upper(), i + 1)
        else:
            backtrack(sub + s[i], i + 1)

    backtrack("", 0)
    return res
